code/Path_Distance_TableGenerator.py

```diff
@@ -0,0 +1,311 @@
+
+    # Find nodes matching the start and end city/state pairs
+    start_nodes = [node[0] for node in nodes if node[1] == start_city and
+                   (node[2] == start_state or not start_state) and
+                   node[3] == start_country]
+    end_nodes = [node[0] for node in nodes if node[1] == end_city and
+                 (node[2] == end_state or not end_state) and
+                 node[3] == end_country]
+
+    print('# of start nodes:', len(start_nodes), file=sys.stderr)
+    print('# of end nodes:', len(end_nodes), file=sys.stderr)
+
+    def _node_str(n):
+        return f'{n["city"]}/{n["state"]}/{n["country"]}'
+
+    # Find the shortest path for each combination of start and end nodes
+    all_paths = []
+    for start_node in start_nodes:
+        for end_node in end_nodes:
+            try:
+                shortest_path = nx.shortest_path(G, source=start_node, target=end_node)
+                shortest_path_location = [_node_str(G.nodes[node]) for node in shortest_path]
+                all_paths.append(shortest_path_location)
+            except NetworkXNoPath:
+                pass
+                #print(f"No path found from {start_city}, {start_country} to {end_city}, {end_country}")
+
+    # Print the result
+    #print(f"Shortest paths from {start_city}, {start_country} to {end_city}, {end_country}:")
+    path_distribution = defaultdict(int)
+    for path in all_paths:
+        path_distribution[str(path)] += 1
+    
+    table_rows = []
+
+    print(f"Path distribution from {start_city}, {start_state}, {start_country} to {end_city},{end_state}, {end_country}:", file=sys.stderr)
+    for path, count in path_distribution.items():
+        path = transform_location_format(path)
+        path = ast.literal_eval(path)
+        
+        print(f"{count}: {path}", file=sys.stderr)
+        
+        print("Calculating the whole distance for the upper path", file=sys.stderr)
+        distance_sum = 0
+        distance_list = []
+        for i in range(1, len(path)):
+            single_distance = get_distance_between_two_router(path[i-1], path[i])
+            distance_list.append(single_distance)
+            distance_sum += single_distance
+        print("the distance for the whole path is", distance_sum, "it is consist of", distance_list, file=sys.stderr)
+        # Join items within each list
+        path_list = '|'.join(map(str, path))
+        distance_list = '|'.join(map(str, distance_list))
+
+        # Combine both strings into one table row
+        table_row = str(count) + '\t' + str(distance_sum) + '\t' + path_list + '\t' + distance_list
+
+        table_rows.append(table_row)
+    table = '\n'.join(table_rows)
+
+    print(table)
+    # Close the database connection
+    conn.close()
+
+
+if __name__ == "__main__":
+    # Parse command-line arguments
+    src, dst = parse_arguments()
+
+    initialize_the_global_graph('../database/test.db')
+
+    find_shortest_path('../database/test.db',
```

Comment:
Can we use a more formal name? Let's use `igdb.db` .

Reply:
fixed, actually, "python3 iGDB.py -c igdb.db" is required before using it. I will put it into the comment of the script.